Charts

martin-ebay opened the pull request:
<!-- Insert GitHub issue number below -->
Fixes #1909

<!-- Select which type of PR this is -->
- [x] This PR contains CSS changes
- [ ] This PR does not contain CSS changes

## Description
adding a handful of tokens for light and dark for data viz colors, as well as updating magenta 3 to match the design team brand colors.

added the donut-chart and chart-legend including documentation.

## Notes
The new tokens will support a set of ebayui-core chart components that are still in the works.

## Screenshots
<!-- Upload screenshots of UI before & after these changes -->

## Checklist
<!-- Acknowledge completion of steps in checklists below. Delete lists that are not applicable -->

<!-- For all PR types -->
- [x] I verify the build is in a non-broken state
- [x] I verify all changes are within scope of the linked issue

<!-- For CSS changes -->
- [x] I regenerated all CSS files under dist folder
- [x] I tested the UI in all supported browsers
- [ ] I did a visual regression check of the components impacted by doing a Percy build and approved the build
- [x] I tested the UI in dark mode and RTL mode
- [x] I added/updated/removed Storybook coverage as appropriate

ArtBlue commented on src/less/donut-chart/donut-chart.less:
```diff
@@ -0,0 +1,193 @@
+    transform: translate(100%, 50%);
+}
+.donut-chart__tooltip.donut-chart__tooltip--top .donut-chart__tooltip-container::after {
+    border-bottom: 10px solid var(--color-neutral-0);
```
Same thing here - simplify.

ArtBlue commented on src/less/donut-chart/stories/donut-chart.stories.js:
```diff
@@ -0,0 +1,55 @@
+export default { title: 'Donut-chart' };
+
+export const basic = () => `
```
The colors look off to me. 
<img width="527" alt="image" src="https://user-images.githubusercontent.com/1675667/199601827-9feb3f0e-e2aa-4960-8589-dee1d90da925.png">

ArtBlue commented on src/less/donut-chart/stories/donut-chart.stories.js:
```diff
@@ -0,0 +1,55 @@
+</div>
+`;
+
+export const patterFill = () => `
```
Patterns look to be missing. 
<img width="530" alt="image" src="https://user-images.githubusercontent.com/1675667/199601942-a3a15a45-07b3-47d3-8fb2-438adac3ceaa.png">

ArtBlue commented on docs/_includes/donut-chart.html:
```diff
@@ -0,0 +1,156 @@
+<div id="donut-chart">
+    {% include section-header.html name="donut-chart" version=page.versions.donut-chart %}
+    <p>The donut chart is used to show the breakdown of a single part-to-whole relationship within a total.</p>
+    <h3>Default Chart Legend</h3>
```
The headings and copy look to be related to the legend. Maybe it was a copy/paste issue? They should be referencing the donut chart.

martin-ebay replied:
totally a copy paste issue. fixing it up

martin-ebay commented on dist/svg/icon/icon-star-empty.svg:
```diff
@@ -1,5 +0,0 @@
-<svg viewbox="0 0 22 22" xmlns="http://www.w3.org/2000/svg">
```
I was trying to figure that out as well, something about the build isn't adding it.  I'll dig some more

martin-ebay replied:
@ArtBlue so the reason these star svg icons are gone is because in the src svg folder there are now two versions for each a -small and -large.  Guess a previous build didn't remove them before

martin-ebay commented on src/less/chart-legend/chart-legend.less:
```diff
@@ -0,0 +1,111 @@
+}
+.chart-legend--hidden {
+    border: 0;
+    clip: rect(0, 0, 0, 0);
```
I used this class to hide the display of the percentage in the legend from the screen but allow a screen reader to see it.  @ArtBlue do we have a utility class for doing this already?

ArtBlue replied:
@martin-ebay , I see. I could be wrong, but I believe that's what `.clipped()` does.

agliga replied:
We have a `.clipped` class from util
https://github.com/eBay/skin/blob/master/dist/mixins/utility-mixins.less#L1-L10

martin-ebay replied:
👍

martin-ebay commented on src/less/chart-legend/chart-legend.less:
```diff
@@ -0,0 +1,111 @@
+}
+.chart-legend__item--quaternary::before {
+    background: repeating-linear-gradient(
+        30deg,
```
just the 30deg part, or the whole repeating-linear-gradient background value?

agliga replied:
I would do the whole thing, especially if we can reuse it in the file
See https://github.com/eBay/skin/blob/master/src/less/progress-stepper/progress-stepper.less#L7-L16
And 
https://github.com/eBay/skin/blob/master/src/less/progress-stepper/progress-stepper.less#L149-L153

It's much cleaner this way.

martin-ebay commented on src/less/chart-legend/chart-legend.less:
```diff
@@ -0,0 +1,111 @@
+}
+.chart-legend ul li::before {
+    border-radius: 2px;
+    content: "";
```
Yes this is needed to render the stylized bullets for the list.  Without the blank content the before pseudo element will not be added to the dom.

martin-ebay commented on src/less/chart-legend/chart-legend.less:
```diff
@@ -0,0 +1,111 @@
+    width: 8px;
+}
+.chart-legend ul li::after {
+    content: "";
```
same as above.  This after pseudo element is used to make the entire list item capture mouse events without gaps between the bullet, text, and value text elements

martin-ebay commented on src/less/chart-legend/stories/chart-legend.stories.js:
```diff
@@ -0,0 +1,61 @@
+            <span class="chart-legend--hidden">33.30%</span>
+        </li>
+        <li id="chart-legend-label-2" data-id="2" class="chart-legend__item--septenary">
+            Value 3
```
Fixing this with updated stories, forgot to update them when updating these components

martin-ebay commented on src/less/chart-legend/stories/chart-legend.stories.js:
```diff
@@ -0,0 +1,61 @@
+        <li id="chart-legend-label-2" data-id="2" class="chart-legend__item--septenary">
+            Value 3
+            <span class="value">$ 1</span>
+            <span class="chart-legend--hidden">16.67%</span>
```
Fixing this with updated stories, forgot to update them when updating these components

martin-ebay commented on src/less/chart-legend/stories/chart-legend.stories.js:
```diff
@@ -0,0 +1,61 @@
+</div>
+`;
+
+export const patternfill = () => `
```
removing patternfill as it is no longer a thing

martin-ebay commented on src/less/donut-chart/donut-chart.less:
```diff
@@ -0,0 +1,193 @@
+.donut-chart__tooltip {
+    color: var(--color-neutral-7);
+    display: block;
+    font-size: 12px;
```
for the font size or the color?

agliga replied:
We have font sizes defined here: https://github.com/eBay/skin/blob/master/src/less/variables/variables.less#L23-L31

agliga commented on dist/donut-chart/donut-chart.css:
```diff
@@ -0,0 +1,184 @@
+  transition: opacity 0.3s;
+}
+.donut-chart__segment--primary {
+  fill: var(--data-viz-primary, var(--color-data-viz-chart-primary));
+  stroke: var(--data-viz-primary, var(--color-data-viz-chart-primary));
```
So generally the way we handle variables is this:

The first one is going to be a token variable, which is per component. This allows a user to override colors for this component only. The second one is fine, it's the one used in the tokens file.
The first variable should not be used in the code at all. This is part of a future project

So I would change this as:
`fill: var(--donut-chart-color-primary, var(--etc))`
Or something like that. See https://github.com/eBay/skin/blob/master/src/less/button/button.less#L98


I would also add, lets use a mixin for this, or add one if there isn't (I see fill has one, but stroke does not)
https://github.com/eBay/skin/blob/master/src/less/mixins/private/token-mixins.less#L81

agliga commented on src/less/donut-chart/donut-chart.less:
```diff
@@ -0,0 +1,185 @@
+    transition: opacity 0.3s;
+}
+.donut-chart__segment--primary {
+    fill: var(--data-viz-primary, var(--color-data-viz-chart-primary));
+    stroke: var(--data-viz-primary, var(--color-data-viz-chart-primary));
```
So generally the way we handle variables is this:

The first one is going to be a token variable, which is per component. This allows a user to override colors for this component only. The second one is fine, it's the one used in the tokens file.
The first variable should not be used in the code at all. This is part of a future project

So I would change this as:
fill: var(--donut-chart-color-primary, var(--etc))
Or something like that. See https://github.com/eBay/skin/blob/master/src/less/button/button.less#L98

I would also add, lets use a mixin for this, or add one if there isn't (I see fill has one, but stroke does not)
https://github.com/eBay/skin/blob/master/src/less/mixins/private/token-mixins.less#L81

agliga commented on src/less/donut-chart/donut-chart.less:
```diff
@@ -0,0 +1,185 @@
+    transform: translate(-50%, 100%);
+}
+.donut-chart__tooltip.donut-chart__tooltip--left .donut-chart__tooltip-container::after {
+    border: 10px solid transparent;
+    border-right: 10px solid var(--color-neutral-0);
+    bottom: 50%;
+    content: "";
+    display: block;
+    left: 1px;
+    position: absolute;
+    transform: translate(-100%, 50%);
+}
+.donut-chart__tooltip.donut-chart__tooltip--right .donut-chart__tooltip-container::after {
+    border: 10px solid transparent;
+    border-left: 10px solid var(--color-neutral-0);
+    bottom: 50%;
+    content: "";
+    display: block;
+    position: absolute;
+    right: 1px;
+    transform: translate(100%, 50%);
```
I see a lot of repeated classes here
If we use `BEM` properly, we don't need to repeat these classes.
For the tooltip we can do `class="donut-chart__tooltip donut-chart__tooltip--top` and it should get both class definitions
For BEM, we also shouldn't need to have nested selectors, keep it simple like:
`.donut-chart__tooltip-container::after {`
`.donut-chart__tooltip--top .donut-chart__tooltip-container::after {`


(This might be what Arthur is referring to the simplify)

agliga commented on src/less/donut-chart/donut-chart.less:
```diff
@@ -0,0 +1,185 @@
+    display: block;
+    font-size: 30px;
+    font-weight: 700;
+    line-height: 40px;
```
We try to stay away from using line heights. If you use the `font-size` variable it should be sufficient.

agliga commented on src/less/donut-chart/donut-chart.less:
```diff
@@ -0,0 +1,185 @@
+    margin-right: 30px;
+    max-width: 190px;
+}
+.donut-chart__legend-right .donut-chart-legend__container {
+    max-width: 210px;
+    width: 50%;
+}
+.donut-chart__legend-left .donut-chart-legend__container {
+    max-width: 210px;
+    width: 50%;
+}
```
Can we just do:

```
.donut-chart-legend__container {
   max-width: 210px;
   width: 50%
}
```

IMO, we should set all base variables and have the left/right ones just either `unset` variables or override them.

agliga commented on src/less/chart-legend/chart-legend.less:
```diff
@@ -0,0 +1,111 @@
+    width: 8px;
+}
+.chart-legend__item--primary::before {
+    background-color: var(--data-viz-primary, var(--color-data-viz-chart-primary));
```
Use `.background-color-token` mixin here.

agliga commented on src/less/chart-legend/chart-legend.less:
```diff
@@ -0,0 +1,111 @@
+}
+.chart-legend__item--primary::before {
+    background-color: var(--data-viz-primary, var(--color-data-viz-chart-primary));
+    border: 1px solid var(--data-viz-primary, var(--color-data-viz-chart-primary));
```
Use `.border-color-token` mixin here.

agliga commented on dist/tokens/evo-dark.css:
```diff
@@ -44,5 +44,27 @@
+        --color-data-viz-chart-quaternary-stroke: var(--color-teal-3);
+        --color-data-viz-chart-quinary-background: var(--color-teal-7);
+        --color-data-viz-chart-quinary-stroke: var(--color-teal-3);
+        --color-data-viz-tooltip-shadow-primary: #ffffff26;
+        --color-data-viz-tooltip-shadow-secondary: #ffffff2b;
```
We should remove these. They should not be needed. (See comment below on the drop shadow for tooltip)

agliga commented on src/less/donut-chart/donut-chart.less:
```diff
@@ -0,0 +1,185 @@
+    background: var(--color-neutral-0);
+    border-radius: 8px;
+    display: block;
+    filter: drop-shadow(0 2px 7px var(--color-data-viz-tooltip-shadow-primary))
+        drop-shadow(0 5px 7px var(--color-data-viz-tooltip-shadow-secondary));
```
We should reuse https://github.com/eBay/skin/blob/master/src/less/mixins/private/bubble-mixins.less#L4-L5

agliga commented on src/less/donut-chart/donut-chart.less:
```diff
@@ -0,0 +1,185 @@
+.donut-chart__tooltip__value {
+    display: block;
+}
+.donut-chart__tooltip-container {
```
Lets reuse this if we can: https://github.com/eBay/skin/blob/master/src/less/tooltip/tooltip.less#L13-L20

agliga reviewed (commented):
First pass. I think we mainly need to switch to reusable code. A lot of this code we already have in other parts. We should reuse it here, rather than rewriting it. 
This will help us in the future maintainance

agliga commented on src/less/donut-graph/donut-graph.less:
```diff
@@ -0,0 +1,84 @@
+@import "../variables/variables.less";
+@import "../mixins/private/token-mixins.less";
+@import "../mixins/private/bubble-mixins.less";
+@import "../tooltip/tooltip.less";
```
This isn't needed. It will duplicate a bunch of code
Instead add it to here so it generates the correct dependencies: https://github.com/eBay/skin/blob/master/scripts/generate-imports/config.json#L23
Something like:
`'donut-graph': ['tooltip']`

martin-ebay replied:
@agliga if I remove the import line 63 of this file throws an error as I'm using a tooltip class like a mixin and spreading it's styles into the .donut-graph__tooltip-container class.

agliga commented on src/less/donut-graph/donut-graph.less:
```diff
@@ -0,0 +1,84 @@
+    opacity: @inactive-opacity;
+}
+
+.donut-graph__tooltip {
```
Is this needed if we are using tooltip?

martin-ebay replied:
@agliga yes this is needed, not all of the code was reusable, as it caused double display and double position values in the compiled code which fails the linter.  I also had to prevent pointer-events since this tooltip follows the mouse and it caused a flicker when it accepted pointer events.

agliga reviewed (commented):
* Please target `15.2.0`
* There's one comment about making a gradient as a variable. Please address that
Had some questions/comments about tooltip

ianmcburnie commented:
I'll need to do a check over with kb & screen reader...

martin-ebay commented:
@ianmcburnie for the skin changes?

ianmcburnie commented:
@martin-ebay Yeah, for the HTML/ARIA structure/semantics

martin-ebay commented:
@ianmcburnie cool, keep in mind that the donut-graph / chart have some help from JS for keyboard focus management in ebayui

ianmcburnie commented:
A few quick comments on donut-chart to keep things moving...

1. both the root div and the paths have `aria-labelledby` attributes that do not resolve to any `id`, which leaves them in an unlabeled state. Feel free to switch them out for aria-label for the sake of this standalone version (i.e. a version without a corresponding legend).
2. I see `tabindex=-1` on the paths which suggests a roving-tab index pattern. We should implement a working roving tabindex on this example using makeup-roving-tabindex.
3. If we have a roving tabindex on the paths, then the root div no longer needs to be keyboard focusable. Tabindex="0" can be removed from the root div.

martin-ebay commented:
@ianmcburnie would it work to just change the tabindex"-1" values to tabindex="0" so that the user tabs through them a single time, or do we want it to loop back?

martin-ebay commented:
@ianmcburnie I've pushed updates, I used tabindex="0" on the paths and removed it from the containers so you can tab through things.  I also changed the path aria-labelledby to aria-label values for the donut-graph, I left the labelled by for the donut-chart as they should resolve the ids

ianmcburnie commented:
Also, I’m not totally convinced we need the full donut + legend + title as a combined module in skin… or if we do do it, maybe as a generic “chart-layout” module to be agnostic of what chart goes inside.

martin-ebay commented:
The other charts would not work with the chart-layout as it only works with our custom svg donut-graph and charity-legend.  Do you want me to move all this CSS for the donut-chart / chart-layout into ebayui-core?

ianmcburnie commented:
> I used tabindex="0" on the paths and removed it from the containers so you can tab through things.

That would force a user to tab through every slice. We want them to be able to easily TAB past the widget and only interact with slices if they want to, using arrow keys (similar behavior as a radio group or tabs). 

A line like this should do it:

`RovingTabindex.createLinear(widgetEl, 'path', { wrap: true });`

Here's an example:
https://github.com/eBay/skin/blob/master/docs/src/js/main.js#L267

More info:
https://ebay.gitbook.io/mindpatterns/techniques/roving-tabindex
https://github.com/makeup/makeup-js/tree/master/packages/makeup-roving-tabindex
https://makeup.github.io/makeup-js/makeup-roving-tabindex/index.html

ianmcburnie commented:
> Do you want me to move all this CSS for the donut-chart / chart-layout into ebayui-core?

Let's leave it as is for now. Although I do see opportunity to simplify and fix some of the BEM.

This isn't valid BEM:

```css
<div aria-labelledby="donutcharttitle" class="donut-chart donut-chart__legend-horizontal donut-chart__legend-left"
```

Perhaps starting with the simplest use case first (stacked, 3 segments), which would require only a single block class ('donut-chart') and then build from there?

I also have a couple of questions/concerns about the legend.

1. We should probably be using a description list
2. I'm not sure we need two versions. The "table" version should suffice for both cases (converted to dl, dt, dd) and then leave it up to the developer what they want to add "clipped" to.

martin-ebay commented:
@ianmcburnie I've pushed updated with the roving tab index as well as changing the chart legend to use the definition list, as well as removed the "minimal" vs "table" displays and defaulted to table. 

I still left the span with the clipped class on the percentage for now as we have no designs that show the percentage.
